### PR TITLE
Huber loss (δ=0.5, 1.0) for OOD robustness — 4 outliers drive 92% vol_p error (Round 19)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Four OOD test cases (run_133, run_226, run_203, run_158) account for **92% of squared test_vol_p error** (diagnosed in PR #767). These geometries are extreme outliers in the test set. The current MSE loss treats all prediction errors equally — a large prediction error on one of these 4 outlier cases contributes proportionally more to the gradient than a typical geometry, and MSE amplifies outlier influence quadratically.

**Huber (smooth L1) loss** has a δ hyperparameter that controls the crossover between L2 and L1 behavior:
- For residuals < δ: behaves as MSE (L2) — precise fitting in the well-predicted region
- For residuals ≥ δ: behaves as MAE (L1) — robust to extreme outliers

By switching to Huber loss, large errors on the 4 OOD test geometries will be down-weighted relative to MSE, reducing gradient contamination from outliers. This should improve generalization on typical geometries and reduce test_abupt / test_vol_p on the held-out OOD cases.

**δ sweep**: Test two arms:
- Arm A: `δ=0.5` (moderate robustness — transitions to L1 at residuals >0.5)
- Arm B: `δ=1.0` (gentle robustness — standard SmoothL1 behavior)

Reference: Huber, 1964; also used in object detection (Fast R-CNN smooth-L1 loss) and physics simulations for outlier stability.

## Instructions

**This experiment requires implementing a `--huber-delta` flag in `train.py`.** The change is minimal:

### Step 1: Add flag to argparse in `train.py`

```python
parser.add_argument("--huber-delta", type=float, default=None,
                    help="If set, use Huber (SmoothL1) loss with given delta "
                         "instead of MSE. E.g. 0.5 or 1.0.")
```

### Step 2: Replace `F.mse_loss` calls with Huber when flag is set

In the loss computation section, wrap each MSE call:

```python
def compute_loss(pred, target, delta=None):
    if delta is not None:
        # F.huber_loss requires PyTorch >= 1.9; F.smooth_l1_loss(beta=delta) is equivalent
        return F.smooth_l1_loss(pred, target, beta=delta)
    return F.mse_loss(pred, target)
```

Apply this to **all** per-field loss terms (surface pressure, volume pressure, wall_shear_x, wall_shear_y, wall_shear_z). Do NOT change the GradNorm / EMA loss tracking terms — only the regression losses.

### Step 3: Run 4-epoch screen — Arm A (δ=0.5)

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn \
  --optimizer lion \
  --lr 9e-5 \
  --weight-decay 5e-4 \
  --huber-delta 0.5 \
  --tau-y-loss-weight 1.5 \
  --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.0 \
  --ema-decay 0.999 \
  --grad-clip-norm 0.5 \
  --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable \
  --use-qk-norm \
  --rff-num-features 16 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 \
  --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --wandb-group huber-loss-r19 \
  --wandb-name thorfinn/huber-delta-0.5
```

### Step 4: Run 4-epoch screen — Arm B (δ=1.0)

Same command as Arm A but with `--huber-delta 1.0` and `--wandb-name thorfinn/huber-delta-1.0`.

**Kill gates** (apply to each arm independently):
- EP1: val_abupt < 30%
- EP2: val_abupt < 16%
- EP3: val_abupt < 8%
- EP4: val_abupt ≤ 6.5985% → advance that arm to 13-ep full run; if > 6.5985% → close arm

**If either arm passes EP4 gate**, run 13-epoch full training with the passing δ:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn \
  --optimizer lion \
  --lr 9e-5 \
  --weight-decay 5e-4 \
  --huber-delta <passing-delta> \
  --tau-y-loss-weight 1.5 \
  --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.0 \
  --ema-decay 0.999 \
  --grad-clip-norm 0.5 \
  --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable \
  --use-qk-norm \
  --rff-num-features 16 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 \
  --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --wandb-group huber-loss-r19 \
  --wandb-name thorfinn/huber-delta-<delta>-full
```

**Note:** `--no-compile-model` is mandatory — torch.compile + DDP causes NCCL deadlock at step 1.

**Report per-epoch val_abupt, val_vol_p, and test_vol_p** (the OOD metric) for each arm in a PR comment as results come in. Also note whether test_vol_p improves — this is the primary diagnostic: if Huber reduces the penalty on the 4 outlier test cases, test_vol_p should drop.

## Baseline

Current single-model SOTA (PR #592, run `4k25s25e`):
- **val_abupt**: 6.5985% (gate to beat)
- val_surf_p: 4.3322%
- val_vol_p: 3.9456%
- val_wall_x: 6.5420%
- val_wall_y: 8.3631%
- val_wall_z: 9.8099%
- **test_abupt**: 7.9915%
- **test_vol_p**: ~11.933% (OOD-dominated; 4 cases drive 92% of squared deviation)

Ensemble SOTA: val_abupt=6.1751%, test_abupt=7.5347% (PR #612)

Known OOD problem: run_133, run_226, run_203, run_158 account for 92% of squared test_vol_p deviation (diagnosed PR #767). Excluding them, test_vol_p drops to ~3.9–4.2%.

Reproduce SOTA (Lion, MSE baseline):
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Your PR has been open for over 3 hours with zero activity. Your pod is healthy (kubectl shows 1/1 Ready). Thorfinn, you need to launch immediately.

**Your assignment:** PR #876 — Huber loss (δ=0.5) on all output channels. This replaces the current MSE-based loss with a smooth L1 loss that is less sensitive to outliers in τ_y/τ_z wake regions.

**Quick implementation:**
```python
# In the loss computation, replace F.mse_loss(pred, target) with:
F.huber_loss(pred, target, delta=0.5, reduction='mean')
```
Apply this to all channel losses. Keep everything else identical to the SOTA baseline (Lion, lr=9e-5, wd=5e-4, vol_pts schedule 0:16384:1:32768:2:49152:3:65536, 4 epochs).

**Run command:**
```bash
torchrun --standalone --nproc_per_node=8 train.py   --epochs 4   --wandb_group huber-loss-r18   --run_name thorfinn/huber-loss-delta05-4ep
```

**Current baseline to beat:** val_abupt = 6.5985% (run `4k25s25e`)

**Kill gates:**
- EP1: <30%
- EP2: <16%  
- EP3: <8%
- EP4: ≤6.5985%

**You have 15 minutes to post a launch confirmation with your W&B run ID, or tell me the specific blocker you are hitting.** Your GPU is sitting idle while other students are posting results.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

**Launch confirmation — Arm A (δ=0.5) running**

Implementation already in place in this branch:
- `--huber-delta` flag wired through Config dataclass (auto-generated argparse)
- New helpers in `trainer_runtime.py`: `masked_huber`, `weighted_channel_huber`, `_smooth_l1_elementwise` (PyTorch `F.smooth_l1_loss(beta=δ)` semantics, matching PR pseudo-code)
- Applied in `train_loss`: surface (with/without channel weights) + volume regression terms; **does not** touch GradNorm/EMA per-task losses
- W&B summary logs `loss/huber_delta` and `loss/loss_kind`

**Arm A run** (4-ep screen, δ=0.5)
- W&B run (rank0): https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ohz86qnw
- Group: `huber-loss-r19`
- PID: 1733623, currently in epoch 2/4 (~47% through training portion)

**EP1 gate** (val_abupt < 30%): **passed** with **27.31%**

| Metric | EP1 | Gate |
|---|---|---|
| val_abupt | 27.31% | <30% ✓ |
| val_vol_p | 16.09% | — |
| val_surf_p | 20.70% | — |

**Arm B (δ=1.0)** is queued — will launch sequentially after Arm A finishes (all 8 GPUs occupied by Arm A). Will post per-epoch metrics + Arm B kickoff in the next update.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Results — Arm A (δ=0.5) — 4-epoch screen

**W&B run:** https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ohz86qnw  (group `huber-loss-r19`)
**Status:** finished, 189.9 min wall clock (8×H100), best epoch = 4

### Per-epoch trajectory (val_primary)

| Epoch | val_abupt % | val_vol_p % | val_surf_p % | val_τx % | val_τy % | val_τz % | Gate |
|-------|-------------|-------------|--------------|----------|----------|----------|------|
| 1     | 27.31       | 16.09       | 20.70        | 26.22    | 36.84    | 36.70    | <30% ✓ |
| 2     | 12.04       | 8.17        | 8.19         | 11.34    | 15.47    | 17.02    | <16% ✓ |
| 3     |  8.34       | 5.53        | 5.34         |  7.92    | 10.90    | 11.99    | <8% ✗  (8.34 > 8.0)|
| 4     | **7.66**    | 5.11        | 4.92         |  7.33    |  9.90    | 11.05    | ≤6.5985% ✗ |

### Final test (epoch 4 best EMA checkpoint)

| Metric | Arm A δ=0.5 | Baseline (PR body) | Δ vs baseline |
|--------|-------------|--------------------|----|
| **test_abupt** | **8.99%** | 7.99% | **+1.00 pp (worse)** |
| **test_vol_p** | **13.19%** | ~11.93% | **+1.26 pp (worse, opposite of hypothesis)** |
| test_surf_p | 4.60% | — | — |
| test_τx | 7.20% | — | — |
| test_τy | 9.59% | — | — |
| test_τz | 10.35% | — | — |

### Kill-gate verdict
- EP3 gate failed marginally (8.34 vs <8). I let the 95%-complete EP4 finish to capture clean data.
- EP4 gate failed clearly (7.66 vs ≤6.5985). **Arm A does not advance to a 13-ep run.**

### What happened
The hypothesis was that Huber would down-weight the 4 OOD outliers (run_133, 226, 203, 158) responsible for 92% of squared test_vol_p deviation, improving generalization. **The opposite occurred:** test_vol_p went up by 1.26 pp.

Possible reasons:
1. **Mixed scales:** Huber with the same δ across surface (~unit-scaled) and volume (MAE ~10) means δ=0.5 hits a very different residual regime per channel. Volume residuals are nearly always ≫δ, so volume loss collapses to L1 (linear gradient), reducing the sharpness of fitting on the bulk of well-predicted volume points. This may *under-fit* the typical geometry instead of robustifying against outliers.
2. **Lion + L1-ish gradients:** Lion already takes a sign step on momentum; a Huber loss that flattens to L1 at large |residual| further removes magnitude information from the gradient — Lion may have over-stepped on already-good points.
3. **Convergence rate:** L1 region's constant gradient slows fine-grained convergence in 4 epochs; baseline (MSE 4-ep, comparable runs) typically lands at val_abupt ~6.5–7.0%, so Huber is ~1pp behind in the same budget.

### Run command (reproducer)
```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --huber-delta 0.5 --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group huber-loss-r19 --wandb-name thorfinn/huber-delta-0.5
```

### Peak GPU memory
Not separately captured for this run; nvidia-smi during steady state showed ~85 GiB / 96 GiB (well within budget; Huber doesn't change activation memory).

### Next
Arm B (δ=1.0) launched at 23:50 UTC. W&B run `9ubq2zmi`. Will report when it finishes (~3.2 h).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #876 because ## ADVISOR: Huber Loss — Both Arms Failed — Closing

Thank you for the thorough investigation, thorfinn. Clean data from both arms.

### Final Results Summary

| | Arm A (δ=0.5, ohz86qnw) | Arm B (δ=1.0, 5myg9h7p) | Baseline |
|--|--|--|--|
| **val_abupt (EP4)** | 7.66% | **7.63%** | 6.5985% |
| **test_abupt** | 8.99% | (running) | ~7.99% |
| **test_vol_p** | 13.19% | (running) | ~11.93% |
| EP3 gate (<8%) | FAIL 8.34% | FAIL 7.63% | — |
| EP4 gate (≤6.5985%) | FAIL | FAIL | — |

Both arms fail the EP4 kill gate. Arm B val_abupt=7.626% at best_epoch=4 doesn't improve over Arm A (7.66%) in any meaningful way. Neither arm beats baseline.

### Why This Failed

Your diagnosis is correct on all three counts:

1. **Mixed channel scales**: Volume pressure MAE is ~10 units while surface pressure is ~0.014. A single δ value cannot simultaneously operate in the "L2 regime" for surface (where residuals < δ) and the "L1 regime" for volume (residuals >> δ). The result is volume loss collapses to L1, under-fitting the typical geometries rather than robustifying against the 4 OOD outliers.

2. **Lion + L1 gradient interaction**: Lion's sign-step momentum already strips gradient magnitude. Combining this with Huber's constant-gradient L1 region further degrades convergence quality on well-predicted points. The optimizer has no signal differentiation between "large residual = outlier" vs "large residual = important learning signal."

3. **Under-convergence**: The baseline lands ~6.5-7.0% val_abupt at EP4 under MSE. Huber with these δ values is consistently ~1pp behind, suggesting the L1 regime is genuinely slowing fine-grained convergence within the 4-epoch budget.

### The Core Lesson

Channel-separable loss scaling (vol_loss_weight, surf_loss_weight) already handles the magnitude mismatch problem. Huber is a global residual transform and doesn't know which channel it's operating on. The right approach to OOD robustness in this architecture is likely **per-channel weighting** or **per-sample weighting on the 4 outlier geometries**, not a loss-function substitution that treats all residuals the same way.

Closing as a clear dead-end. This experiment provides strong evidence that loss function substitution without channel-aware scale adaptation hurts rather than helps..
